lib/wallet/src/Cardano/Wallet/Write/UTxOAssumptions.hs

```diff
@@ -0,0 +1,57 @@
+{-# LANGUAGE LambdaCase #-}
+module Cardano.Wallet.Write.UTxOAssumptions
+    ( UTxOAssumptions (..)
+    , assumedInputScriptTemplate
+    , assumedTxWitnessTag
+    , addressSatisfiesAssumptions
+    )
+    where
+
+import Prelude
+
+import Cardano.Ledger.Shelley.API
+    ( Addr (..), Credential (..) )
+import Cardano.Wallet.TxWitnessTag
+    ( TxWitnessTag (..) )
+import Cardano.Wallet.Write.Tx
+    ( Address )
+
+import qualified Cardano.Address.Script as CA
+import qualified Cardano.Wallet.Primitive.Types.Address as W
+
+-- | Assumptions about the UTxO which are needed for coin-selection.
+data UTxOAssumptions
+    = AllKeyPaymentCredentials
+    -- ^ Assumes all 'UTxO' entries have addresses with the post-Shelley
+    -- key payment credentials.
+    | AllByronKeyPaymentCredentials
+    -- ^ Assumes all 'UTxO' entries have addresses with the boostrap/byron
+    -- key payment credentials.
+    | AllScriptPaymentCredentialsFrom
+    -- ^ Assumes all 'UTxO' entries have addresses with script
+    -- payment credentials, where the scripts are both derived
+    -- from the 'ScriptTemplate' and can be looked up using the given function.
+        !CA.ScriptTemplate
+        !(W.Address -> CA.Script CA.KeyHash)
+
+assumedInputScriptTemplate :: UTxOAssumptions -> Maybe CA.ScriptTemplate
+assumedInputScriptTemplate = \case
+    AllKeyPaymentCredentials -> Nothing
+    AllByronKeyPaymentCredentials -> Nothing
+    AllScriptPaymentCredentialsFrom scriptTemplate _ -> Just scriptTemplate
+
+assumedTxWitnessTag :: UTxOAssumptions -> TxWitnessTag
+assumedTxWitnessTag = \case
+    AllKeyPaymentCredentials -> TxWitnessShelleyUTxO
+    AllByronKeyPaymentCredentials -> TxWitnessByronUTxO
+    AllScriptPaymentCredentialsFrom {} -> TxWitnessShelleyUTxO
+
+addressSatisfiesAssumptions :: UTxOAssumptions -> Address -> Bool
+addressSatisfiesAssumptions = go
+  where
+    -- 'go' helper saves horizontal space
+    go AllKeyPaymentCredentials          (Addr _ KeyHashObj{}    _) = True
+    go AllScriptPaymentCredentialsFrom{} (Addr _ ScriptHashObj{} _) = True
+    go AllByronKeyPaymentCredentials     (AddrBootstrap _)          = True
+    go _                                 _                          = False
```

Comment:
Though does it actually save horizontal space? 🤔 
```suggestion
addressSatisfiesAssumptions ua a = case (ua, a) of
    (AllKeyPaymentCredentials         , (Addr _ KeyHashObj{}    _)) = True
    (AllScriptPaymentCredentialsFrom{}, (Addr _ ScriptHashObj{} _)) = True
    (AllByronKeyPaymentCredentials    , (AddrBootstrap _)         ) = True
    _                                                               = False
```

Reply:
Compared to `validateAddress AllKeyPaymentCredentials (Addr`: yes. Compared to what you have above: no, however I think the lack of `( ... , ...)` noise makes the logic clearer. I renamed `go` to `valid` though. Does this sound fine to you?